README


This archive contains the distribution Email-Folder,
version 0.859:

  read all the messages from a folder as Email::Simple objects

This software is copyright (c) 2006 by Simon Wistow.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


This README file was generated by Dist::Zilla::Plugin::Readme v5.020.



Changes
Revision history for Email-Folder

0.859     2014-09-16 18:29:32-04:00 America/New_York
        - all the changes in this release contributed by Pali Rohár
        - support for mboxcl and mboxcl2
        - add ability to read from a filehandle
        - add next_from and next_messageref methods

0.858     2014-01-07 10:54:18-05:00 America/New_York
          typo fixes in docs (thanks, Florian Schlichting)

0.857     2013-08-05 18:16:44 America/New_York
          eliminate extra Pod tests

0.856     2013-08-05 18:12:09 America/New_York
          avoid duplicating headers when data is malformed (#31132)

          make MH folder reader work on VMS

0.855     2009-07-26
          fix pod tests

0.854     2007-12-23
          fix a bug in which the last header of a mbox message could be lost

0.853     2007-03-23
          packaging improvements

0.852     2006-08-22
          README

0.851     2006-08-01
          use ExtUtils::MakeMaker, to standardize PEP
          add pod coverage tests and improve coverage

0.85      Tuesday 4th April, 2006
          Fix clobbering of $_ in Mbox.pm
          Put FH in binmode to fix non-ascii mails on Windows

0.84      Friday 17th December, 2004
          Added in ability for subclasses of Email::Folder::Mbox to override
          IO::File creation.

          Added in information about Perl Email Project

HEAD
          Added Email::Folder::MH by Ricardo SIGNES

0.83      Sunday 9th May, 2004
          Rerelease without a broken manifest (rt.cpan.org $6049)

0.82      Monday 22nd March, 2003
          Add Email::Folder::Ezmlm to the dist.

          Switch to a trad Makefile.PL

0.81
          Slow down the release numbering a bit.  I'd like to be stable at
          the point of 1.0 or so.

          Added ->reader as an accessor to the Email::Reader subclass

          Added ->tell to ::Mbox, and seek_to as an option

0.8       29th May, 2003
          mboxcl2 handling is now a little more paranoid, and will recover from
          bad values

          Mbox.pm now expects the From_ line to match either of
              /^From (?:Mon|Tue|Wed|Thu|Fri|Sat|Sun)/
              /^From \S+\s+(?:Mon|Tue|Wed|Thu|Fri|Sat|Sun)/
          (based on mutt mbox.c/from.c)

          Document that ->messages is a one-shot operation.

0.7       Tuesday 27th May, 2003
          Mbox.pm can grok mboxcl2

          INCOMPATIBLE CHANGE: Email::Folder::Mbox now takes line endings as
          an explicit option rather than trying to guess.

0.6       Sunday 11th May, 2003
          Rename the files in the test maildir so we can unpack in Win32 systems
          Release with a auto-generated Makefile.PL, for CPAN.pm users

0.5       Tuesday 29th April, 2003
          Add iterator interface, plus move the backends to be classes

0.4       Thursday 24th April, 2003
          Add line-ending detection to Email::Folder::MBox
          temporarily disable signing, as it's confusing

0.3       Tuesday 22nd April, 2003
          Just another release of 0.2, now that CPAN repects my authoritay!

0.2       Tuesday 22nd April, 2003
          Lose the (hidden) dependency on UNIVERSAL::require

0.1       Initial version
          Very rough idea


lib/Email/Folder.pm
use strict;
use warnings;
package Email::Folder;
# ABSTRACT: read all the messages from a folder as Email::Simple objects
$Email::Folder::VERSION = '0.859';
use Carp;
use Email::Simple;
use Email::FolderType 0.6 qw/folder_type/;

#pod =head1 SYNOPSIS
#pod
#pod  use Email::Folder;
#pod
#pod  my $folder = Email::Folder->new("some_file");
#pod
#pod  print join "\n", map { $_->header("Subject") } $folder->messages;
#pod
#pod =head1 METHODS
#pod
#pod =head2 new($folder, %options)
#pod
#pod Takes the name of a folder, and a hash of options
#pod
#pod If a 'reader' option is passed in then that is
#pod used as the class to read in messages with.
#pod
#pod =cut

sub new {
    my $class  = shift;
    my $folder = shift || carp "Must provide a folder name\n";
    my %self = @_;

    my $reader;

    if ($self{reader}) {
        $reader = $self{reader};
    } else {
        $reader = "Email::Folder::".folder_type($folder);
    }
    eval "require $reader" or die $@;

    $self{_folder} = $reader->new($folder, @_);

    return bless \%self, $class;
}

#pod =head2 messages
#pod
#pod Returns a list containing all of the messages in the folder.  Can only
#pod be called once as it drains the iterator.
#pod
#pod =cut

sub messages {
    my $self = shift;

    my @messages = $self->{_folder}->messages;
    my @ret;
    while (my $body = shift @messages) {
        push @ret, $self->bless_message( $body );
    }
    return @ret;
}


#pod =head2 next_message
#pod
#pod acts as an iterator.  reads the next message from a folder.  returns
#pod false at the end of the folder
#pod
#pod =cut

sub next_message {
    my $self = shift;

    my $body = $self->{_folder}->next_message or return;
    $self->bless_message( $body );
}


#pod =head2 bless_message($message)
#pod
#pod Takes a raw RFC822 message and blesses it into a class.
#pod
#pod By default this is an Email::Simple object but can easily be overridden
#pod in a subclass.
#pod
#pod For example, this simple subclass just returns the raw rfc822 messages,
#pod and exposes the speed of the parser.
#pod
#pod  package Email::RawFolder;
#pod  use base 'Email::Folder';
#pod  sub bless_message { $_[1] };
#pod  1;
#pod
#pod =cut

sub bless_message {
    my $self    = shift;
    my $message = shift || die "You must pass a message\n";

    return Email::Simple->new($message);
}


#pod =head2 reader
#pod
#pod read-only accessor to the underlying Email::Reader subclass instance
#pod
#pod =cut

sub reader {
    my $self = shift;
    return $self->{_folder};
}

1;

#pod =head1 SEE ALSO
#pod
#pod L<Email::LocalDelivery>, L<Email::FolderType>, L<Email::Simple>
#pod
#pod =cut

__END__

=pod

=encoding UTF-8

=head1 NAME

Email::Folder - read all the messages from a folder as Email::Simple objects

=head1 VERSION

version 0.859

=head1 SYNOPSIS

 use Email::Folder;

 my $folder = Email::Folder->new("some_file");

 print join "\n", map { $_->header("Subject") } $folder->messages;

=head1 METHODS

=head2 new($folder, %options)

Takes the name of a folder, and a hash of options

If a 'reader' option is passed in then that is
used as the class to read in messages with.

=head2 messages

Returns a list containing all of the messages in the folder.  Can only
be called once as it drains the iterator.

=head2 next_message

acts as an iterator.  reads the next message from a folder.  returns
false at the end of the folder

=head2 bless_message($message)

Takes a raw RFC822 message and blesses it into a class.

By default this is an Email::Simple object but can easily be overridden
in a subclass.

For example, this simple subclass just returns the raw rfc822 messages,
and exposes the speed of the parser.

 package Email::RawFolder;
 use base 'Email::Folder';
 sub bless_message { $_[1] };
 1;

=head2 reader

read-only accessor to the underlying Email::Reader subclass instance

=head1 SEE ALSO

L<Email::LocalDelivery>, L<Email::FolderType>, L<Email::Simple>

=head1 AUTHORS

=over 4

=item *

Simon Wistow <simon@thegestalt.org>

=item *

Richard Clamp <richardc@unixbeard.net>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2006 by Simon Wistow.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Email/Folder/MH.pm
use strict;
use warnings;
package Email::Folder::MH;
# ABSTRACT: reads raw RFC822 mails from an mh folder
$Email::Folder::MH::VERSION = '0.859';
use Carp;
use IO::File;
use Email::Folder::Reader;
use parent 'Email::Folder::Reader';

#pod =head1 SYNOPSIS
#pod
#pod This isa Email::Folder::Reader - read about its API there.
#pod
#pod =head1 DESCRIPTION
#pod
#pod It's yet another email folder reader!  It reads MH folders.
#pod
#pod =cut

sub _what_is_there {
    my $self = shift;
    my $dir = $self->{_file};

    croak "$dir does not exist"     unless (-e $dir);
    croak "$dir is not a directory" unless (-d $dir);

    my @messages;
                opendir(DIR,"$dir") or croak "Could not open '$dir'";
                foreach my $file (readdir DIR) {
                    if ($^O eq 'VMS'){
                        next unless $file =~ /\A\d+\.\Z/;
                    } else {
                        next unless $file =~ /\A\d+\Z/;
                    }
                    push @messages, "$dir/$file";
                }

    $self->{_messages} = \@messages;
}

sub next_message {
    my $self = shift;
    my $what = $self->{_messages} || $self->_what_is_there;

    my $file = shift @$what or return;
    local *FILE;
    open FILE, $file or croak "couldn't open '$file' for reading";
    join '', <FILE>;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Email::Folder::MH - reads raw RFC822 mails from an mh folder

=head1 VERSION

version 0.859

=head1 SYNOPSIS

This isa Email::Folder::Reader - read about its API there.

=head1 DESCRIPTION

It's yet another email folder reader!  It reads MH folders.

=head1 AUTHORS

=over 4

=item *

Simon Wistow <simon@thegestalt.org>

=item *

Richard Clamp <richardc@unixbeard.net>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2006 by Simon Wistow.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Email/Folder/Mbox.pm
use strict;
use warnings;
package Email::Folder::Mbox;
# ABSTRACT: reads raw RFC822 mails from an mbox file
$Email::Folder::Mbox::VERSION = '0.859';
use Carp;
use IO::File;
use Email::Folder::Reader;
use parent 'Email::Folder::Reader';

#pod =head1 SYNOPSIS
#pod
#pod This isa Email::Folder::Reader - read about its API there.
#pod
#pod =head1 DESCRIPTION
#pod
#pod Does exactly what it says on the tin - fetches raw RFC822 mails from an
#pod mbox.
#pod
#pod The mbox format is described at http://www.qmail.org/man/man5/mbox.html
#pod
#pod We attempt to read an mbox as through it's the mboxcl2 variant,
#pod falling back to regular mbox mode if there is no C<Content-Length>
#pod header to be found.
#pod
#pod =head2 OPTIONS
#pod
#pod The new constructor takes extra options.
#pod
#pod =over
#pod
#pod =item C<fh>
#pod
#pod When filename is set to C<"FH"> than Email::Folder::Mbox will read mbox
#pod archive from filehandle C<fh> instead from disk file C<filename>.
#pod
#pod =item C<eol>
#pod
#pod This indicates what the line-ending style is to be.  The default is
#pod C<"\n">, but for handling files with mac line-endings you would want
#pod to specify C<eol =E<gt> "\x0d">
#pod
#pod =item C<jwz_From_>
#pod
#pod The value is taken as a boolean that governs what is used match as a
#pod message separator.
#pod
#pod If false we use the mutt style
#pod
#pod  /^From \S+\s+(?:Mon|Tue|Wed|Thu|Fri|Sat|Sun)/
#pod  /^From (?:Mon|Tue|Wed|Thu|Fri|Sat|Sun)/;
#pod
#pod If true we use
#pod
#pod  /^From /
#pod
#pod In deference to this extract from L<http://www.jwz.org/doc/content-length.html>
#pod
#pod   Essentially the only safe way to parse that file format is to
#pod   consider all lines which begin with the characters ``From ''
#pod   (From-space), which are preceded by a blank line or
#pod   beginning-of-file, to be the division between messages.  That is, the
#pod   delimiter is "\n\nFrom .*\n" except for the very first message in the
#pod   file, where it is "^From .*\n".
#pod
#pod   Some people will tell you that you should do stricter parsing on
#pod   those lines: check for user names and dates and so on.  They are
#pod   wrong.  The random crap that has traditionally been dumped into that
#pod   line is without bound; comparing the first five characters is the
#pod   only safe and portable thing to do. Usually, but not always, the next
#pod   token on the line after ``From '' will be a user-id, or email
#pod   address, or UUCP path, and usually the next thing on the line will be
#pod   a date specification, in some format, and usually there's nothing
#pod   after that.  But you can't rely on any of this.
#pod
#pod Defaults to false.
#pod
#pod =item C<unescape>
#pod
#pod This boolean value indicates whenever lines which starts with
#pod
#pod  /^>+From /
#pod
#pod should be unescaped (= removed leading '>' char). This is needed for
#pod mboxrd and mboxcl variants. But there is no way to detect for used mbox
#pod variant, so default value is false.
#pod
#pod =item C<seek_to>
#pod
#pod Seek to an offset when opening the mbox.  When used in combination with
#pod ->tell you may be able to resume reading, with a trailing wind.
#pod
#pod =item <next_message>
#pod
#pod This returns next message as string
#pod
#pod =item <next_messageref>
#pod
#pod This returns next message as ref to string
#pod
#pod =item C<tell>
#pod
#pod This returns the current filehandle position in the mbox.
#pod
#pod =item C<next_from>
#pod
#pod This returns the From_ line for next message. Call it before ->next_message.
#pod
#pod =back
#pod
#pod =cut

sub defaults {
    ( eol => "\n")
}

sub _open_it {
    my $self = shift;
    my $file = $self->{_file};
    my $fh = $self->{fh};

    unless ($file eq "FH" and $fh) {
        # sanity checking
        croak "$file does not exist" unless (-e $file);
        croak "$file is not a file"  unless (-f $file);

        local $/ = $self->{eol};
        $fh = $self->_get_fh($file);
    }

    if (seek $fh, tell($fh), 0) {
        # Enable using seek only if $fh is seekable
        $self->{seekable} = 1;
    } else {
        # Otherwise use cache for simulating backward seeks
        $self->{cache} = [];
    }

    if ($self->{seek_to}) {
        unless ($self->{seekable}) {
            croak "$file is not seekable but seek_to was set";
        }
        # we were told to seek.  hope it all goes well
        seek $fh, $self->{seek_to}, 0;
    }
    else {
        local $/ = $self->{eol};
        my $firstline = <$fh>;
        if ($firstline) {
            croak "$file is not an mbox file" unless $firstline =~ /^From /;
        }
        $self->{from} = $firstline;
    }

    $self->{_fh} = $fh;
}

sub _get_fh {
    my $self = shift;
    my $file = shift;
    my $fh = IO::File->new($file) or croak "Cannot open $file";
    binmode($fh);
    return $fh;
}

sub _read_nextline {
    my $self = shift;
    if (not $self->{seekable} and @{$self->{cache}}) {
        return shift @{$self->{cache}};
    }
    my $fh = $self->{_fh};
    return <$fh>;
}

use constant debug => 0;
my $count;

sub next_from {
    my $self = shift;
    $self->_open_it unless $self->{_fh};
    return $self->{from};
}

sub next_messageref {
    my $self = shift;

    my $fh = $self->{_fh} || $self->_open_it;
    local $/ = $self->{eol};

    my $mail = '';
    my $prev = '';
    my $last;
    my $inheaders = 1;
    ++$count;
    print "$count starting scanning at line $.\n" if debug;

    while (my $line = _read_nextline($self)) {
        if ($line eq $/ && $inheaders) { # end of headers
            print "$count end of headers at line $.\n" if debug;
            $inheaders = 0; # stop looking for the end of headers
            my $pos; # where to go back to if it goes wrong
            $pos = tell $fh if $self->{seekable};

            # look for a content length header, and try to use that
            if ($mail =~ m/^Content-Length:\s*(\d+)$/mi) {
                my @cache;
                $mail .= $prev;
                $prev = '';
                my $length = $1;
                print " Content-Length: $length\n" if debug;
                my $read = '';
                while (my $bodyline = _read_nextline($self)) {
                    push @cache, $bodyline unless $self->{seekable};
                    last if length $read >= $length;
                    # unescape From_
                    $bodyline =~ s/^>(>*From )/$1/ if $self->{unescape};
                    $read .= $bodyline;
                }
                # grab the next line (should be /^From / or undef)
                my $next = _read_nextline($self);
                if (!defined $next || $next =~ /^From /) {
                    $self->{from} = $next;
                    $mail .= "$/$read";
                    return \$mail;
                }
                push @cache, $next unless $self->{seekable};
                # seek back and scan line-by-line like the header
                # wasn't here
                print " Content-Length assertion failed '$next'\n" if debug;
                if ($self->{seekable}) {
                    seek $fh, $pos, 0;
                }
                else {
                    unshift @{$self->{cache}}, @cache;
                }
            }

            # much the same, but with Lines:
            if ($mail =~ m/^Lines:\s*(\d+)$/mi) {
                my @cache;
                $mail .= $prev;
                $prev = '';
                my $lines = $1;
                print " Lines: $lines\n" if debug;
                my $read = '';
                for (1 .. $lines) {
                    my $bodyline = _read_nextline($self);
                    last unless defined $bodyline;
                    push @cache, $bodyline unless $self->{seekable};
                    # unescape From_
                    $bodyline =~ s/^>(>*From )/$1/ if $self->{unescape};
                    $read .= $bodyline;
                }
                my $ign = _read_nextline($self); # trailing newline
                my $next = _read_nextline($self);
                if (!defined $next || $next =~ /^From /) {
                    $self->{from} = $next;
                    $mail .= "$/$read";
                    return \$mail;
                }
                push @cache, $ign, $next unless $self->{seekable};
                # seek back and scan line-by-line like the header
                # wasn't here
                print " Lines assertion failed '$next'\n" if debug;
                if ($self->{seekable}) {
                    seek $fh, $pos, 0;
                }
                else {
                    unshift @{$self->{cache}}, @cache;
                }
            }
        }

        if ($prev eq $/ && ($line =~ $self->_from_line_re)) {
            $mail .= $prev;
            $last = $line;
            last;
        }

        $mail .= $prev;
        $prev = $line;

        # unescape From_
        $prev =~ s/^>(>*From )/$1/ if $self->{unescape};
    }
    print "$count end of message line $.\n" if debug;
    $self->{from} = $last;
    return unless $mail;
    return \$mail;
}

sub next_message {
    my $self = shift;
    my $ref = $self->next_messageref;
    return unless $ref;
    return ${$ref};
}

my @FROM_RE;
BEGIN {
  @FROM_RE = (
    # according to mutt:
    #   A valid message separator looks like:
    #   From [ <return-path> ] <weekday> <month> <day> <time> [ <tz> ] <year>
    qr/^From (?:\S+\s+)?(?:Mon|Tue|Wed|Thu|Fri|Sat|Sun)/,

    # though, as jwz rants, only this is reliable and portable
    qr/^From /,
  );
}

sub _from_line_re {
  return $FROM_RE[ $_[0]->{jwz_From_} ? 1 : 0 ];
}

sub tell {
    my $self = shift;
    return tell $self->{_fh};
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Email::Folder::Mbox - reads raw RFC822 mails from an mbox file

=head1 VERSION

version 0.859

=head1 SYNOPSIS

This isa Email::Folder::Reader - read about its API there.

=head1 DESCRIPTION

Does exactly what it says on the tin - fetches raw RFC822 mails from an
mbox.

The mbox format is described at http://www.qmail.org/man/man5/mbox.html

We attempt to read an mbox as through it's the mboxcl2 variant,
falling back to regular mbox mode if there is no C<Content-Length>
header to be found.

=head2 OPTIONS

The new constructor takes extra options.

=over

=item C<fh>

When filename is set to C<"FH"> than Email::Folder::Mbox will read mbox
archive from filehandle C<fh> instead from disk file C<filename>.

=item C<eol>

This indicates what the line-ending style is to be.  The default is
C<"\n">, but for handling files with mac line-endings you would want
to specify C<eol =E<gt> "\x0d">

=item C<jwz_From_>

The value is taken as a boolean that governs what is used match as a
message separator.

If false we use the mutt style

 /^From \S+\s+(?:Mon|Tue|Wed|Thu|Fri|Sat|Sun)/
 /^From (?:Mon|Tue|Wed|Thu|Fri|Sat|Sun)/;

If true we use

 /^From /

In deference to this extract from L<http://www.jwz.org/doc/content-length.html>

  Essentially the only safe way to parse that file format is to
  consider all lines which begin with the characters ``From ''
  (From-space), which are preceded by a blank line or
  beginning-of-file, to be the division between messages.  That is, the
  delimiter is "\n\nFrom .*\n" except for the very first message in the
  file, where it is "^From .*\n".

  Some people will tell you that you should do stricter parsing on
  those lines: check for user names and dates and so on.  They are
  wrong.  The random crap that has traditionally been dumped into that
  line is without bound; comparing the first five characters is the
  only safe and portable thing to do. Usually, but not always, the next
  token on the line after ``From '' will be a user-id, or email
  address, or UUCP path, and usually the next thing on the line will be
  a date specification, in some format, and usually there's nothing
  after that.  But you can't rely on any of this.

Defaults to false.

=item C<unescape>

This boolean value indicates whenever lines which starts with

 /^>+From /

should be unescaped (= removed leading '>' char). This is needed for
mboxrd and mboxcl variants. But there is no way to detect for used mbox
variant, so default value is false.

=item C<seek_to>

Seek to an offset when opening the mbox.  When used in combination with
->tell you may be able to resume reading, with a trailing wind.

=item <next_message>

This returns next message as string

=item <next_messageref>

This returns next message as ref to string

=item C<tell>

This returns the current filehandle position in the mbox.

=item C<next_from>

This returns the From_ line for next message. Call it before ->next_message.

=back

=head1 AUTHORS

=over 4

=item *

Simon Wistow <simon@thegestalt.org>

=item *

Richard Clamp <richardc@unixbeard.net>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2006 by Simon Wistow.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Email/Folder/Ezmlm.pm
use strict;
use warnings;
package Email::Folder::Ezmlm;
# ABSTRACT: reads raw RFC822 mails from an ezmlm archive
$Email::Folder::Ezmlm::VERSION = '0.859';
use Carp;
use Email::Folder::Maildir;
use parent 'Email::Folder::Maildir';

# we're a little complicit, just redefining an internal method, but
# that's fine, we maintain both piles :)

sub _what_is_there {
    my $self = shift;
    my $dir = $self->{_file};

    croak "$dir does not exist"           unless (-e $dir);
    croak "$dir is not an ezmlm archive"  unless (-d $dir);
    croak "$dir is not an ezmlm archive"  unless (-e "$dir/archive" && -d "$dir/archive");

    my $num;
    if (my $fh = IO::File->new("$dir/num")) {
        ($num) = (<$fh> =~ m/^(\d+)/);
    }

    $self->{_messages} = [ map {
        sprintf '%s/archive/%d/%02d', $dir, int $_ / 100, $_ % 100
    } 1..$num ];
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Email::Folder::Ezmlm - reads raw RFC822 mails from an ezmlm archive

=head1 VERSION

version 0.859

=head1 AUTHORS

=over 4

=item *

Simon Wistow <simon@thegestalt.org>

=item *

Richard Clamp <richardc@unixbeard.net>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2006 by Simon Wistow.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Email/Folder/Reader.pm
use strict;
use warnings;
package Email::Folder::Reader;
# ABSTRACT: reads raw RFC822 mails from a box
$Email::Folder::Reader::VERSION = '0.859';
use Carp;

#pod =head1 SYNOPSIS
#pod
#pod  use Email::Folder::Reader;
#pod  my $box = Email::Folder::Reader->new('somebox');
#pod  print $box->messages;
#pod
#pod or, as an iterator
#pod
#pod  use Email::Folder::Reader;
#pod  my $box = Email::Folder::Reader->new('somebox');
#pod  while ( my $mail = $box->next_message ) {
#pod      print $mail;
#pod  }
#pod
#pod =head1 METHODS
#pod
#pod =head2 new($filename, %options)
#pod
#pod your standard class-method constructor
#pod
#pod =cut

sub new {
    my $class = shift;
    my $file  = shift || croak "You must pass a filename";
    bless { eval { $class->defaults },
            @_,
            _file => $file }, $class;
}

#pod =head2 ->next_message
#pod
#pod returns the next message from the box, or false if there are no more
#pod
#pod =cut

sub next_message {
}

#pod =head2 ->messages
#pod
#pod Returns all the messages in a box
#pod
#pod =cut

sub messages {
    my $self = shift;

    my @messages;
    while (my $message = $self->next_message) {
        push @messages, $message;
    }
    return @messages;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Email::Folder::Reader - reads raw RFC822 mails from a box

=head1 VERSION

version 0.859

=head1 SYNOPSIS

 use Email::Folder::Reader;
 my $box = Email::Folder::Reader->new('somebox');
 print $box->messages;

or, as an iterator

 use Email::Folder::Reader;
 my $box = Email::Folder::Reader->new('somebox');
 while ( my $mail = $box->next_message ) {
     print $mail;
 }

=head1 METHODS

=head2 new($filename, %options)

your standard class-method constructor

=head2 ->next_message

returns the next message from the box, or false if there are no more

=head2 ->messages

Returns all the messages in a box

=head1 AUTHORS

=over 4

=item *

Simon Wistow <simon@thegestalt.org>

=item *

Richard Clamp <richardc@unixbeard.net>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2006 by Simon Wistow.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


lib/Email/Folder/Maildir.pm
use strict;
use warnings;
package Email::Folder::Maildir;
# ABSTRACT: reads raw RFC822 mails from a maildir
$Email::Folder::Maildir::VERSION = '0.859';
use Carp;
use IO::File;
use Email::Folder::Reader;
use parent 'Email::Folder::Reader';

#pod =head1 SYNOPSIS
#pod
#pod This isa Email::Folder::Reader - read about its API there.
#pod
#pod =head1 DESCRIPTION
#pod
#pod Does exactly what it says on the tin - fetches raw RFC822 mails from a
#pod maildir.
#pod
#pod The maildir format is described at
#pod L<http://www.qmail.org/man/man5/maildir.html>
#pod
#pod =cut

sub _what_is_there {
    my $self = shift;
    my $dir = $self->{_file};

    croak "$dir does not exist"    unless (-e $dir);
    croak "$dir is not a maildir"  unless (-d $dir);
    croak "$dir is not a maildir"  unless (-e "$dir/cur" && -d "$dir/cur");
    croak "$dir is not a maildir"  unless (-e "$dir/new" && -d "$dir/new");

    my @messages;
    # ignore the tmp directory although the spec
    # says to delete anything in tmp/ that is older than 36 hours
    for my $sub (qw(new cur)) {
        opendir(DIR,"$dir/$sub") or croak "Could not open '$dir/$sub'";
        foreach my $file (readdir DIR) {
            next if $file =~ /^\./; # as suggested by DJB
            push @messages, "$dir/$sub/$file";
        }
    }

    $self->{_messages} = \@messages;
}

sub next_message {
    my $self = shift;
    my $what = $self->{_messages} || $self->_what_is_there;

    my $file = shift @$what or return;
    local *FILE;
    open FILE, $file or croak "couldn't open '$file' for reading";
    join '', <FILE>;
}

1;

__END__

=pod

=encoding UTF-8

=head1 NAME

Email::Folder::Maildir - reads raw RFC822 mails from a maildir

=head1 VERSION

version 0.859

=head1 SYNOPSIS

This isa Email::Folder::Reader - read about its API there.

=head1 DESCRIPTION

Does exactly what it says on the tin - fetches raw RFC822 mails from a
maildir.

The maildir format is described at
L<http://www.qmail.org/man/man5/maildir.html>

=head1 AUTHORS

=over 4

=item *

Simon Wistow <simon@thegestalt.org>

=item *

Richard Clamp <richardc@unixbeard.net>

=back

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2006 by Simon Wistow.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.020.
use strict;
use warnings;



use ExtUtils::MakeMaker ;



my %WriteMakefileArgs = (
  "ABSTRACT" => "read all the messages from a folder as Email::Simple objects",
  "AUTHOR" => "Simon Wistow <simon\@thegestalt.org>, Richard Clamp <richardc\@unixbeard.net>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "Email-Folder",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "Email::Folder",
  "PREREQ_PM" => {
    "Carp" => 0,
    "Email::FolderType" => "0.6",
    "Email::Simple" => 0,
    "IO::File" => 0,
    "constant" => 0,
    "parent" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "TEST_REQUIRES" => {
    "Test::More" => "0.96"
  },
  "VERSION" => "0.859",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Carp" => 0,
  "Email::FolderType" => "0.6",
  "Email::Simple" => 0,
  "IO::File" => 0,
  "Test::More" => "0.96",
  "constant" => 0,
  "parent" => 0,
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



